### PR TITLE
Add delegation commands to CLI

### LINKED ISSUE #1348: CLI: add delegation related commands/features

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Adds create, delay, and destroy delegation commands to the wallet CLI. Also cleans up some CLI command types.

## Links to any relevant issues

Closes #1348

### CODE REVIEW

File: cli/src/wallet_cli/mod.rs [hunk 1]
```diff
@@ -66,19 +66,19 @@ pub enum WalletCommand {
     /// Burn an amount of native token.
     BurnNativeToken {
         /// Token ID to be burnt, e.g. 0x087d205988b733d97fb145ae340e27a8b19554d1ceee64574d7e5ff66c45f69e7a0100000000.
-        token_id: String,
+        token_id: TokenId,
         /// Amount to be burnt, e.g. 100.
-        amount: String,
+        amount: U256,
```
Comment (Thoralf-M) [inline cli/src/wallet_cli/mod.rs]: Is this really working as expected? If I remember correctly by default it didn't use `::from_dec_str()` by default but something else like hex or so which then ended up in wrong values

Comment (Thoralf-M) [reply cli/src/wallet_cli/mod.rs]: Btw would be nice to have an own PR for such changes since this is not related to delegation

File: sdk/src/client/api/block_builder/input_selection/mod.rs [hunk 2]
```diff
@@ -379,7 +380,12 @@ impl InputSelection {
     pub fn select(mut self) -> Result<Selected, Error> {
         if !OUTPUT_COUNT_RANGE.contains(&(self.outputs.len() as u16)) {
             // If burn or mana allotments are provided, outputs will be added later.
-            if !(self.outputs.is_empty() && (self.burn.is_some() || self.mana_allotments != 0)) {
+            if !(self.outputs.is_empty()
+                && (self.burn.is_some()
+                    || self.mana_allotments != 0
+                    // TODO: ??? is this okay ???
+                    || self.available_inputs.iter().any(|o| o.output.is_delegation())))
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/input_selection/mod.rs]: The delegation should be added as required_input, right? For the consolidation and claiming PRs (still draft) I changed it to `!self.required_inputs.is_empty()` which I think is better because then you know for sure that these output will be part of the inputs

File: sdk/src/client/api/block_builder/input_selection/mod.rs [hunk 3]
```diff
@@ -393,29 +399,56 @@ impl InputSelection {
...
-            return Err(Error::InvalidOutputCount(self.outputs.len()));
+            self.outputs.extend(storage_deposit_returns);
+            self.outputs.extend(rem.iter().map(|r| r.output.clone()));
+
+            // Evaluate the new output requirements
+            self.outputs_requirements(catchall.as_ref().map(|c| &c.output));
+            // Need to re-evaluate amount + mana differences since outputs have storage requirements
+            self.requirements.push(Requirement::Amount);
+            self.requirements.push(Requirement::Mana(0));
+            self.requirements.push(Requirement::NativeTokens);
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/input_selection/mod.rs]: Why is this added? This only checks again if the inputs have all the required native tokens for the outputs?

File: sdk/src/client/api/block_builder/input_selection/mod.rs [hunk 4]
```diff
@@ -393,29 +399,56 @@ impl InputSelection {
...
-        if !OUTPUT_COUNT_RANGE.contains(&(self.outputs.len() as u16)) {
-            return Err(Error::InvalidOutputCount(self.outputs.len()));
+            self.outputs.extend(storage_deposit_returns);
+            self.outputs.extend(rem.iter().map(|r| r.output.clone()));
+
+            // Evaluate the new output requirements
+            self.outputs_requirements(catchall.as_ref().map(|c| &c.output));
+            // Need to re-evaluate amount + mana differences since outputs have storage requirements
+            self.requirements.push(Requirement::Amount);
+            self.requirements.push(Requirement::Mana(0));
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/input_selection/mod.rs]: Why is this needed? And isn't `0` wrong? This value should be the mana amount from the allotments, in `fulfill_mana_requirement()` this gets added together with the output mana values

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: removed

File: sdk/src/client/api/block_builder/input_selection/mod.rs (hunk 3 above)
Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: removed

File: sdk/src/client/api/block_builder/input_selection/mod.rs (hunk 2 above)
Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: removed

Comment (Thoralf-M) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: Not even `!self.required_inputs.is_empty()` is needed?

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: I think the burn implies that

File: cli/src/wallet_cli/mod.rs (hunk 1 above)
Comment (DaughterOfMars) [reply cli/src/wallet_cli/mod.rs]: I made one :)

File: sdk/src/client/api/block_builder/input_selection/remainder.rs [hunk 5]
```diff
@@ -205,48 +205,35 @@ fn create_remainder_outputs(
...
+                    .finish_output()?;
+                remaining_amount = remaining_amount.saturating_sub(output.amount());
+                remainder_outputs.push(output);
+            // Otherwise save this one for the catchall
             } else {
-                // Only last output uses amount diff and needs to be validated
-                remainder.verify_storage_deposit(storage_score_parameters)?;
-            };
-            log::debug!(
-                "Created remainder output of amount {}, mana {} and native token {native_token:?} for {remainder_address:?}",
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/input_selection/remainder.rs:234]: I would like to keep the debug logs for remainder outputs

File: sdk/src/client/api/block_builder/input_selection/remainder.rs [hunk 6]
```diff
@@ -205,48 +205,35 @@ fn create_remainder_outputs(
...
-            let remainder = remainder_builder
-                .add_unlock_condition(AddressUnlockCondition::new(remainder_address.clone()))
-                .with_native_token(native_token)
-                .finish_output()?;
-
-            if n + 1 < native_tokens_len {
-                remaining_amount = remaining_amount.saturating_sub(remainder.amount());
+        for native_token in native_tokens.into_iter().rev() {
+            // If we already picked one for the catchall, create a new remainder output with the min amount
+            if catchall_native_token.is_some() {
```
Comment (Thoralf-M) [inline sdk/src/client/api/block_builder/input_selection/remainder.rs]: How about using `.split_last()` I think that's even clearer
```Rust
    if let Some(native_tokens) = native_tokens_diff {
        if let Some((last, nts)) = native_tokens.split_last() {
            // Save this one for the catchall
            catchall_native_token.replace(*last);
            // Create remainder outputs with min amount
            for native_token in nts {
                let output = BasicOutputBuilder::new_with_minimum_amount(storage_score_parameters)
                    .add_unlock_condition(AddressUnlockCondition::new(remainder_address.clone()))
                    .with_native_token(*native_token)
                    .finish_output()?;
                log::debug!(
                    "Created remainder output of amount {}, mana {} and native token {native_token:?} for {remainder_address:?}",
                    output.amount(),
                    output.mana()
                );
                remaining_amount = remaining_amount.saturating_sub(output.amount());
                remainder_outputs.push(output);
            }
        }
    }
    let mut catchall = BasicOutputBuilder::new_with_amount(remaining_amount)
        .with_mana(mana_diff)
        .add_unlock_condition(AddressUnlockCondition::new(remainder_address.clone()));
    if let Some(native_token) = catchall_native_token {
        catchall = catchall.with_native_token(native_token);
    }
    let catchall = catchall.finish_output()?;
    catchall.verify_storage_deposit(storage_score_parameters)?;
    log::debug!(
        "Created remainder output of amount {}, mana {} and native token {:?} for {remainder_address:?}",
        catchall.amount(),
        catchall.mana(),
        catchall.native_token(),
    );
    remainder_outputs.push(catchall);
```

File: sdk/src/client/api/block_builder/input_selection/remainder.rs (hunk 5 above)
Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/input_selection/remainder.rs:234]: oops

File: sdk/src/client/api/block_builder/input_selection/remainder.rs (hunk 6 above)
Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/input_selection/remainder.rs]: not bad!

Comment (DaughterOfMars) [reply sdk/src/client/api/block_builder/input_selection/remainder.rs]: you can just push this if you want

**Review by Thoralf-M — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]
